src/native/eventpipe/ep-buffer-manager.c

```diff
@@ -1209,12 +1213,33 @@ ep_buffer_manager_write_all_buffers_to_file_v4 (
 					// miscategorize it, but that seems unlikely.
 					uint32_t last_read_delta = last_read_sequence_number - thread_sequence_number;
 					if (0 < last_read_delta && last_read_delta < 0x80000000) {
-						ep_rt_thread_sequence_number_map_add_or_replace (ep_sequence_point_get_thread_sequence_numbers_ref (sequence_point), session_state, last_read_sequence_number);
-						if (!exists)
+						if (exists) {
+							ep_rt_thread_sequence_number_map_remove (ep_sequence_point_get_thread_sequence_numbers_ref (sequence_point), session_state);
+						} else {
 							ep_thread_addref (ep_thread_holder_get_thread (ep_thread_session_state_get_thread_holder_ref (session_state)));
+						}
+						ep_rt_thread_sequence_number_map_add (ep_sequence_point_get_thread_sequence_numbers_ref (sequence_point), session_state, last_read_sequence_number);
+					}
+
+					// if a session_state was exhausted during this sequence point, mark it for deletion
+					if (ep_thread_session_state_get_buffer_list (session_state)->head_buffer == NULL) {
+
+						if (ep_rt_volatile_load_uint32_t_without_barrier (ep_thread_get_unregistered_ref (ep_thread_session_state_get_thread (session_state))) > 0) {
```

Comment:
The unregistered field is getting accessed without proper synchronization with the thread getting unregistered assigning value 1 to it but It seems like that's ok since the next time this method is called it will be marked for deletion. Can we add a comment here saying that there is a possible race here and it's ok to have that race because it'll be marked for deletion upon the next time this method is invoked by the streaming thread?

Reply:
Yes, this was on purpose.  It wasn't worth taking the lock if we just had to wait till the next sequence point to mark it.  I'll add a comment saying as much.